### PR TITLE
Add C++ variant for binaries sample

### PR DESCRIPTION
**Pull request by Dastin-Wilski-Mobica**
* file utils were expanded to support reading and writing binaries

### CODE REVIEW

File: samples/core/binaries/main.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ */
+
+// OpenCL SDK includes
+#include <CL/Utils/Utils.hpp>
+#include <CL/SDK/Context.hpp>
+#include <CL/SDK/Options.hpp>
+#include <CL/SDK/CLI.hpp>
+
+// STL includes
+#include <iostream>
```
Comment (bashbaug) [inline samples/core/binaries/main.cpp:24]: The CI builds are failing on Linux due to a missing include:

```suggestion
#include <climits>
#include <iostream>
```

**Review by bashbaug — CHANGES_REQUESTED** [review]

File: lib/src/Utils/File.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        *error = CL_INVALID_VALUE;
+        return std::vector<unsigned char>();
+    }
+}
+
+
+cl::Program::Binaries
+cl::util::read_binary_files(const std::vector<cl::Device>& devices,
+                            const char* const program_file_name,
```
Comment (MathiasMagnus) [inline lib/src/Utils/File.cpp:66]: If program_file_name is only a base file name, perhaps we should indicate that in the variable name. IntelliSense will confuse people with this variable name _and_ no Doxygen summary popping up.

File: lib/src/Utils/File.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    for (const auto& device : devices)
+    {
+        string device_name = device.getInfo<CL_DEVICE_NAME>();
+        string binary_name =
+            string(program_file_name) + "-" + device_name + ".bin";
+
+
+        binaries.push_back(
+            cl::util::read_binary_file(binary_name.c_str(), error));
+        if (*error == CL_INVALID_VALUE) return binaries;
```
Comment (MathiasMagnus) [inline lib/src/Utils/File.cpp:81]: I think `errHandler` is missing here too.

File: lib/src/Utils/File.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        } catch (std::bad_alloc& ex)
+        {
+            (void)ex;
+            detail::errHandler(CL_OUT_OF_RESOURCES, error, "Bad allocation!");
+            return std::vector<unsigned char>();
+        }
+    }
+    else
+    {
+        *error = CL_INVALID_VALUE;
```
[the remark this answers is not in the document]
Comment (MathiasMagnus) [reply lib/src/Utils/File.cpp:58]: Also, the error is never checked for being a `nullptr`. Check before assigning.

File: samples/core/binaries/main.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                      << "Selected device: "
+                      << devices.at(0).getInfo<CL_DEVICE_NAME>() << "\n"
+                      << std::endl;
+        }
+
+        /// Try to read binary
+        cl::Program::Binaries binaries =
+            cl::util::read_binary_files(devices, "Collatz", &error);
+
+        if (error != CL_SUCCESS)
```
Comment (MathiasMagnus) [inline samples/core/binaries/main.cpp:96]: All samples are built with exception handling enabled. If error handling is implemented using `errHandler`, then in the presence of exceptions, trying to read the binaries which aren't on disk yet result in an exception. As such, compiling the binaries should be in a `catch(cl::util::Error&)` clause, so if any of the files did not exist on disk, then we compile from scratch and save.

Comment (MathiasMagnus) [reply samples/core/binaries/main.cpp:96]: As a sidenote, the utility library could expose a kernel caching mechanism of some sort. Either using the current working directory, or some other more sophisticated storage mechanism. Other libraries roll their own kernel cache (clFFT for instance), but this is a recurring issue and could be captured by a dedicated utility. This is nothing to act upon now, but we should capture in an issue.

**Review by MathiasMagnus — CHANGES_REQUESTED** [review]
Nice addition. Some changes are requested to align better with existing code and behavior.

File: samples/core/binaries/main.cpp (hunk 5 above)
Comment (Dastin-Wilski-Mobica) [reply samples/core/binaries/main.cpp]: Thanks for review!

I've tried to address your comments in the following commit.
I hope its better now.

**Review by MathiasMagnus — CHANGES_REQUESTED** [review]
Good changes. Only one error-path seems to omit exception behavior.

File: lib/src/Utils/File.cpp [hunk 6]
```diff
@@ -45,42 +44,45 @@ cl::util::read_binary_file(const char* const filename, cl_int* const error)
...
 
         binaries.push_back(
             cl::util::read_binary_file(binary_name.c_str(), error));
-        if (*error == CL_INVALID_VALUE) return binaries;
+        if (error != nullptr)
+        {
+            if (*error != CL_SUCCESS)
+            {
+                return cl::Program::Binaries();
+            }
```
[the remark this answers is not in the document]
Comment (Dastin-Wilski-Mobica) [reply lib/src/Utils/File.cpp]: done

**Review by MathiasMagnus — APPROVED** [review]
👍🏻

**Review by bashbaug — APPROVED** [review]
LGTM and the Linux builds are fixed - thanks!

**Pull request merged**